Mathlib/LinearAlgebra/RootSystem/Basic.lean

```diff
@@ -0,0 +1,352 @@
+as part of the data.
+
+We have opted for a design in which some of the uniquely-determined data is included: the bijection
+between roots and coroots is (implicitly) included and the coroots are included for root systems.
+Empirically this seems to be by far the most convenient design and by providing extensionality
+lemmas expressing the uniqueness we expect to get the best of both worlds.
+
+A final point is that we require roots and coroots to be injections from a base indexing type `ι`
+rather than subsets of their codomains. This design was chosen to avoid the bijection between roots
+and coroots being a dependently-typed object. A third option would be to have the roots and coroots
+be subsets but to avoid having a dependently-typed bijection by defining it globally with junk value
+`0` outside of the roots and coroots. This would work but lacks the convenient symmetry that the
+chosen design enjoys: by introducing the indexing type `ι`, one does not have to pick a direction
+(`roots → coroots` or `coroots → roots`) for the forward direction of the bijection. Besides,
+providing the user with the additional definitional power to specify an indexing type `ι` is a
+benefit and the junk-value pattern is a cost.
+
+-/
+
+open Set Function
+open Module hiding reflection
+open Submodule (span)
+open AddSubgroup (zmultiples)
+
+noncomputable section
+
+variable (ι R M N : Type*) [Finite ι]
+  [CommRing R] [AddCommGroup M] [Module R M] [AddCommGroup N] [Module R N]
+
+/-- Given two perfectly-paired `R`-modules `M` and `N`, a root pairing with finite indexing set `ι`
+is the data of an `ι`-indexed subset of `M` ("the roots") and an `ι`-indexed subset of `N`
+("the coroots") satisfying the axioms familiar from the traditional theory of root systems / data.
+
+It exists to allow for a convenient unification of the theories of root systems and root data. -/
+structure RootPairing extends PerfectPairing R M N :=
+  root : ι ↪ M
+  coroot : ι ↪ N
+  root_coroot_two : ∀ i, toLin (root i) (coroot i) = 2
+  mapsTo_preReflection_root :
+    ∀ i, MapsTo (preReflection (root i) (toLin.flip (coroot i))) (range root) (range root)
+  mapsTo_preReflection_coroot :
+    ∀ i, MapsTo (preReflection (coroot i) (toLin (root i))) (range coroot) (range coroot)
+
+attribute [nolint docBlame] RootPairing.root
+attribute [nolint docBlame] RootPairing.coroot
+
+/-- A root datum is a root pairing with coefficients in the integers and for which the root and
+coroot spaces are finitely-generated free Abelian groups.
+
+Note that the latter assumptions `[Free ℤ X₁] [Finite ℤ X₁] [Free ℤ X₂] [Finite ℤ X₂]` should be
+supplied as mixins. -/
+abbrev RootDatum (X₁ X₂ : Type*) [AddCommGroup X₁] [AddCommGroup X₂] := RootPairing ι ℤ X₁ X₂
+
+/-- A root system is a root pairing for which the roots span their ambient module.
+
+Note that this is slightly more general than the usual definition in the sense that `N` is not
```

Comment:
Isn't that also what Bourbaki does? (I didn't check.) I think they also take two modules and a perfect pairing, instead of one module and its dual.

Reply:
In Bourbaki a root system is just a subset of a vector space, subject to hypotheses. So there is no perfect pairing;  implicitly it is thus the canonical one between a (finite-dimensional) vector space and its dual.

Of course for root data (which Bourbaki does not cover) people do usually include a perfect pairing as part of the data.

Reply:
Aah, then I completely misremembered how they treat the topic.